File: src/creature.py
import math

from attributes import AttributeSet
from dice import Dice
from game import game
from item import Container, Gold


class Creature:
    def __init__(
        self, sprite, initial_pos, name="Creature", attribute_override=None, debug=False
    ):
        self.sprite = sprite
        self.position = initial_pos

        # Name of the creature.
        self.name = name

        # Creature base attributes
        if attribute_override is not None:
            self.attributes = attribute_override
        else:
            self.attributes = AttributeSet()

        # Debugging dumps (verbose!)
        self.debug = debug

        # Alive or dead?
        self.alive = True

        # Creature's hitpoints. Can be increased by any number of means.
        # Can only be increased past the maximum hitpoints by a buff from a
        # power or armor piece.
        self.hitpoints = 30

        # Creature's maximum hitpoints. Can only be increased by specific
        # powers, or by levelling up.
        self.maxhitpoints = 30

        # Buffs applied to the creature at this point in time.
        self.buffs = []

        # Poisons applied to the creature at this point in time.
        self.poisons = []

        # Current experience points
        self.xp = 0

        # Next level experience points
        self.next_level_xp = 64

        # Current level
        self.level = 1

        # Turn number. Used for buff timeouts and such.
        self.turn = 1

        # How much gold do we have?
        self.gold = 0

        # Current inventory.
        self.inventory = Container(capacity=10)

        # Some creatures will naturally have a higher defense base. This number
        # is immutable - additional defense points come from the defense bonus
        # which is calculated from armor, and from the CON attribute (and to
        # a lesser extent, the DEX attribute)
        self.defenseBase = 5

        # The attack bonus will be calculated as weapons are wielded or powers
        # used that cause the creature to gain or lose attack bonuses. For
        # example, a piece of armor might increase the defensive bonus but be
        # awfully unwieldy, causing the attack bonus to reduce. Minimum of 2.
        self.attackBonus = 2

        # The defense bonus is adjusted when armor is worn or powers used.
        # Minimum of 2.
        self.defenseBonus = 2

        # All wield points for the creature. None means not wielded, otherwise
        # this will be valid Python object holding information about the item.
        # Note that wielding anything will automatically affect the bonuses -
        # the actual inventory and wield points is for the player's benefit.
        self.wieldpoints = {
            "head": None,  # Helmets and such
            "chest": None,  # Armor
            "arms": None,  # Armor
            "hands": None,  # Weapons
            "legs": None,  # Armor
            "feet": None,
        }  # Armor, dexterity modifiers

        self.default_wield = {
            "head": None,  # Helmets and such
            "chest": None,  # Armor
            "arms": None,  # Armor
            "hands": None,  # Weapons
            "legs": None,  # Armor
            "feet": None,
        }

        # Creatures in battle don't do other AI things like move around.
        self.in_battle = False

        # Creatures can be controlled by an AI that needs to think, too.
        self.ai = None

    def writedebug(self, s):
        if self.debug:
            print(s)

    def rollforattrs(self):
        rolls = self.attributes.rollfor()

        dice = Dice()

        # hp = d20 + CON modifier normalized to 14 as the midpoint

        rolls["hp"] = (dice.roll(1, 20) + rolls["con"][0] - 14, "1d20")
        self.hitpoints = self.maxhitpoints = max(
            5, rolls["hp"][0]
        )  # creatures cannot have less than 5 hp to start with

        return rolls

    def roll_mob_attrs(self):
        dice = Dice()
        # scale gold based on HP of the creature
        count = int(math.ceil(self.maxhitpoints / 15))
        if count <= 0:
            count = 1
        self.gold = dice.rollnamed("%dd20" % (count,)) * 3

    def getWeaponDamage(self):
        # Unarmed combat: 1-6 damage.
        if self.wieldpoints["hands"] == None:
            return "1d6"
        else:
            return self.wieldpoints["hands"].damage()

    def getWeaponCriticalRange(self):
        # Unarmed combat is always a 5% chance of a critical
        if self.wieldpoints["hands"] == None:
            return 20
        else:
            return self.wieldpoints["hands"].criticalRange()

    def getWeaponCriticalMultiplier(self):
        # Unarmed combat will always have a 2x damage critical
        if self.wieldpoints["hands"] == None:
            return 2
        else:
            return self.wieldpoints["hands"].criticalMultiplier()

    def wield(self, point, item):
        # Unwield any existing item, and remove bonuses.
        currentItem = self.wieldpoints[point]
        if currentItem is not None:
            self.attackBonus -= currentItem.getAttackBonus()
            self.defenseBonus -= currentItem.getDefenseBonus()

        # Wield the new item
        self.wieldpoints[point] = item

        # Apply bonuses
        if item is not None:
            self.writedebug(
                "Creature '%s' at turn %d wielded '%s' on '%s' for %d Attack Bonus and %d Defense Bonus"
                % (
                    self.name,
                    self.turn,
                    item.name,
                    point,
                    item.getAttackBonus(),
                    item.getDefenseBonus(),
                )
            )

            self.attackBonus += item.getAttackBonus()
            self.defenseBonus += item.getDefenseBonus()

    def atWieldPoint(self, point):
        return self.wieldpoints[point]

    def buff(self, buff):
        self.buffs += [buff]

        buff.setExpiry(self.turn + buff.getLifetime())

        self.writedebug(
            "Creature '%s' at turn %d has had a buff (%s) applied for %d turns, with +%d HP, +%d Attack Bonus, +%d Defense Bonus."
            % (
                self.name,
                self.turn,
                buff.name,
                buff.getLifetime(),
                buff.getHpBuff(),
                buff.getAttackBuff(),
                buff.getDefenseBuff(),
            )
        )

        self.hitpoints += buff.getHpBuff()
        self.attackBonus += buff.getAttackBuff()
        self.defenseBonus += buff.getDefenseBuff()

    def poison(self, poison):
        self.poisons += [poison]

        self.writedebug(
            "Creature '%s' at turn %d has had poison (%s) applied for %d turns, with %d damage per turn."
            % (
                self.name,
                self.turn,
                poison.name,
                poison.getLifetime(),
                poison.damagePerTurn(),
            )
        )

        poison.setExpiry(self.turn + poison.getLifetime())

    def think(self):
        # Don't do anything for dead creatures.
        if not self.alive:
            return

        if self.ai is not None:
            self.ai.think()

        # Check all buffs for expiry
        removeList = []
        for buff in self.buffs:
            if buff.hasExpired(self.turn):
                self.writedebug(
                    "Creature '%s' at turn %d has had buff '%s' expire."
                    % (self.name, self.turn, buff.name)
                )
                removeList += [buff]

        for buff in removeList:
            self.buffs.remove(buff)

            self.hitpoints -= buff.getHpBuff()
            self.attackBonus -= buff.getAttackBuff()
            self.defenseBonus -= buff.getDefenseBuff()

            # When a buff expires it CANNOT kill the creature, only leave them
            # with very few hitpoints.
            if self.hitpoints < 0:
                self.hitpoints = 1

        # Check poisons for expiry
        removeList = []
        for poison in self.poisons:
            if poison.hasExpired(self.turn):
                self.writedebug(
                    "Creature '%s' at turn %d has had poison '%s' expire."
                    % (self.name, self.turn, poison.name)
                )
                removeList += [poison]

            self.hitpoints -= poison.damagePerTurn()

        for poison in removeList:
            self.poisons.remove(poison)

        # Are we dead?
        if self.hitpoints <= 0:
            self.writedebug("Creature: at turn %d, '%s' died." % (self.turn, self.name))
            self.alive = False
            self.hitpoints = 0
        else:
            self.writedebug(
                "Creature: at turn %d, '%s' has %d HP."
                % (self.turn, self.name, self.hitpoints)
            )

        # Turn complete
        self.turn += 1

    def give(self, item):
        if isinstance(item, Gold):
            self.gold += item.value
            return True
        else:
            return self.inventory.add_item(item)

    def give_xp(self, xp):
        self.xp += xp
        if self.xp >= self.next_level_xp:
            self.next_level_xp *= 2.0
            self.level += 1
            self.maxhitpoints *= 2
            self.hitpoints = self.maxhitpoints  # player heals on level-up
            game().log("Welcome to level %d" % (self.level,))

    def describe_wields(self):
        entries = []
        for point in ("hands", "chest", "head", "arms", "legs", "feet"):
            wield = self.wieldpoints[point]
            if wield is not None:
                entries.append(f"<b>{point.title()}</b>\n  {wield.describe()}")

        if not entries:
            return "Wielding and wearing nothing."
        else:
            return "<b>Wielding:</b>\n%s" % ("\n".join(entries),)

    def ai_control(self, ai):
        self.ai = ai

    def ai_release(self):
        self.ai.detach()
        self.ai = None

File: web/src/creature.ts
import {
  CREATURE_GOLD_MULTIPLIER,
  CREATURE_GOLD_SCALER,
  DEFAULT_WEAPON_CRITICAL_MULTIPLIER,
  DEFAULT_WEAPON_CRITICAL_RANGE,
  DEFAULT_WEAPON_DAMAGE_DICE,
  MOB_CRITICAL_MULTIPLIER_MAXIMUM,
  MOB_CRITICAL_RANGE_MINIMUM,
  PLAYER_BASE_ATTACK_BONUS,
  PLAYER_BASE_DEFENSE,
  PLAYER_BASE_DEFENSE_BONUS,
  PLAYER_HP_HEAL_ON_LEVEL_UP,
  PLAYER_HP_PER_LEVEL_MULTIPLIER,
  PLAYER_CONSTITUTION_BONUS,
  PLAYER_INVENTORY_CAPACITY,
  PLAYER_INITIAL_HP,
  PLAYER_XP_FOR_LEVEL_2,
  PLAYER_XP_GOAL_MULTIPLIER,
} from "./constants";
import { AttributeSet, type AttributeName } from "./attributes";
import { Buff, Container, Gold, InstantEffectItem, Poison, Weapon, WieldableItem } from "./items";
import type { Position, WieldSlot } from "./types";
import type { Dice } from "./dice";

const SLOTS: WieldSlot[] = ["head", "chest", "arms", "hands", "legs", "feet"];

export class Creature {
  alive = true;

  hitpoints = PLAYER_INITIAL_HP;

  maxHitpoints = PLAYER_INITIAL_HP;

  buffs: Buff[] = [];

  poisons: Poison[] = [];

  xp = 0;

  nextLevelXp = PLAYER_XP_FOR_LEVEL_2;

  level = 1;

  turn = 1;

  gold = 0;

  inventory = new Container(PLAYER_INVENTORY_CAPACITY);

  defenseBase = PLAYER_BASE_DEFENSE;

  attackBonus = PLAYER_BASE_ATTACK_BONUS;

  defenseBonus = PLAYER_BASE_DEFENSE_BONUS;

  mob = false;

  inBattle = false;

  damageDealtMultiplier = 1;

  damageTakenMultiplier = 1;

  hitpointCapMultiplier = 1;

  unspentStatPoints = 0;

  readonly wieldpoints: Record<WieldSlot, WieldableItem | null> = {
    head: null,
    chest: null,
    arms: null,
    hands: null,
    legs: null,
    feet: null,
  };

  constructor(
    name: string,
    public position: Position,
    readonly attributes: AttributeSet = new AttributeSet(),
  ) {
    this.name = name;
  }

  name: string;

  currentMaxHitpoints(): number {
    return Math.max(1, Math.floor(this.maxHitpoints * this.hitpointCapMultiplier));
  }

  enforceHitpointCap(): void {
    const cap = this.currentMaxHitpoints();
    this.hitpoints = Math.min(cap, this.hitpoints);
  }

  rollMobGold(dice: Dice): void {
    const count = Math.max(1, Math.ceil(this.maxHitpoints / CREATURE_GOLD_SCALER));
    this.gold = Math.floor(dice.rollNamed(`${count}d20`) * CREATURE_GOLD_MULTIPLIER);
  }

  getWeapon(): Weapon | null {
    const weapon = this.wieldpoints.hands;
    return weapon instanceof Weapon ? weapon : null;
  }

  getWeaponDamage(): string {
    const weapon = this.getWeapon();
    if (!weapon) {
      return DEFAULT_WEAPON_DAMAGE_DICE;
    }
    return weapon.damage();
  }

  getWeaponCriticalRange(): number {
    const weapon = this.getWeapon();
    if (!weapon) {
      return DEFAULT_WEAPON_CRITICAL_RANGE;
    }
    const critRange = weapon.criticalRange();
    if (this.mob) {
      return Math.max(MOB_CRITICAL_RANGE_MINIMUM, critRange);
    }
    return critRange;
  }

  getWeaponCriticalMultiplier(): number {
    const weapon = this.getWeapon();
    if (!weapon) {
      return DEFAULT_WEAPON_CRITICAL_MULTIPLIER;
    }
    const critMult = weapon.criticalMultiplier();
    if (this.mob) {
      return Math.min(MOB_CRITICAL_MULTIPLIER_MAXIMUM, critMult);
    }
    return critMult;
  }

  wield(slot: WieldSlot, item: WieldableItem | null): void {
    const current = this.wieldpoints[slot];
    if (current) {
      this.attackBonus -= current.getAttackBonus();
      this.defenseBonus -= current.getDefenseBonus();
    }

    this.wieldpoints[slot] = item;

    if (item) {
      this.attackBonus += item.getAttackBonus();
      this.defenseBonus += item.getDefenseBonus();
    }
  }

  describeWields(): string {
    const lines: string[] = [];
    for (const slot of SLOTS) {
      const item = this.wieldpoints[slot];
      if (item) {
        lines.push(`${slot}: ${item.describe()}`);
      }
    }
    if (lines.length === 0) {
      return "Wielding nothing.";
    }
    return lines.join("\n");
  }

  addBuff(buff: Buff): void {
    this.buffs.push(buff);
    buff.setExpiry(this.turn + buff.lifetime);

    this.hitpoints += buff.hpBuff;
    this.attackBonus += buff.attackBuff;
    this.defenseBonus += buff.defenseBuff;
  }

  addPoison(poison: Poison): void {
    this.poisons.push(poison);
    poison.setExpiry(this.turn + poison.lifetime);
  }

  think(): void {
    if (!this.alive) {
      return;
    }

    const expiredBuffs: Buff[] = [];
    for (const buff of this.buffs) {
      if (buff.hasExpired(this.turn)) {
        expiredBuffs.push(buff);
      }
    }

    for (const buff of expiredBuffs) {
      const idx = this.buffs.findIndex((candidate) => candidate.id === buff.id);
      if (idx >= 0) {
        this.buffs.splice(idx, 1);
      }
      this.hitpoints -= buff.hpBuff;
      this.attackBonus -= buff.attackBuff;
      this.defenseBonus -= buff.defenseBuff;
      if (this.hitpoints < 1) {
        this.hitpoints = 1;
      }
    }

    const expiredPoisons: Poison[] = [];
    for (const poison of this.poisons) {
      if (poison.hasExpired(this.turn)) {
        expiredPoisons.push(poison);
      }
      this.hitpoints -= poison.damagePerTurn;
    }

    for (const poison of expiredPoisons) {
      const idx = this.poisons.findIndex((candidate) => candidate.id === poison.id);
      if (idx >= 0) {
        this.poisons.splice(idx, 1);
      }
    }

    if (this.hitpoints <= 0) {
      this.alive = false;
      this.hitpoints = 0;
    } else {
      this.enforceHitpointCap();
    }

    this.turn += 1;
  }

  give(item: Gold | WieldableItem | InstantEffectItem): boolean {
    if (item instanceof Gold) {
      this.gold += item.amount;
      return true;
    }
    return this.inventory.add(item);
  }

  spendGold(amount: number): boolean {
    if (this.gold < amount) {
      return false;
    }
    this.gold -= amount;
    return true;
  }

  giveXp(xp: number): { leveled: number; pointsAwarded: number } {
    this.xp += xp;

    let leveled = 0;
    while (this.xp >= this.nextLevelXp) {
      this.level += 1;
      leveled += 1;
      this.unspentStatPoints += 1;
      this.nextLevelXp *= PLAYER_XP_GOAL_MULTIPLIER;
      this.maxHitpoints = Math.floor(this.maxHitpoints * PLAYER_HP_PER_LEVEL_MULTIPLIER);
      const heal = Math.floor(this.maxHitpoints * PLAYER_HP_HEAL_ON_LEVEL_UP);
      this.hitpoints = Math.min(this.currentMaxHitpoints(), this.hitpoints + heal);
    }

    this.enforceHitpointCap();
    return { leveled, pointsAwarded: leveled };
  }

  applyInstantItem(item: InstantEffectItem): number {
    const applied = item.apply(this.hitpoints, this.currentMaxHitpoints());
    this.hitpoints = applied.hp;
    return applied.delta;
  }

  grantAttributePoint(points = 1): void {
    this.unspentStatPoints += Math.max(0, Math.floor(points));
  }

  spendAttributePoint(attr: AttributeName): boolean {
    if (this.unspentStatPoints <= 0) {
      return false;
    }

    this.attributes.modify(attr, 1);
    this.unspentStatPoints -= 1;

    if (attr === "con") {
      this.maxHitpoints += PLAYER_CONSTITUTION_BONUS;
      this.hitpoints = Math.min(
        this.currentMaxHitpoints(),
        this.hitpoints + PLAYER_CONSTITUTION_BONUS,
      );
    }

    this.enforceHitpointCap();
    return true;
  }

  bestHealItem(): InstantEffectItem | null {
    const hpNeeded = this.maxHitpoints - this.hitpoints;
    const heals = this.inventory
      .items()
      .filter((item): item is InstantEffectItem => item instanceof InstantEffectItem && item.hpBoost > 0)
      .sort((a, b) => a.hpBoost - b.hpBoost);

    if (heals.length === 0) {
      return null;
    }

    let fallback = heals[0];
    for (const heal of heals) {
      if (heal.hpBoost >= hpNeeded) {
        return heal;
      }
      fallback = heal;
    }

    return fallback;
  }

  healItems(): InstantEffectItem[] {
    return this.inventory
      .items()
      .filter((item): item is InstantEffectItem => item instanceof InstantEffectItem && item.hpBoost > 0);
  }
}
